main-actions/src/main/scala/sbt/compiler/Eval.scala

```diff
@@ -503,25 +505,26 @@ private[sbt] object Eval {
   def bytes(b: Seq[Array[Byte]]): Array[Byte] = bytes(b.length) ++ b.flatten.toArray[Byte]
   def bytes(b: Boolean): Array[Byte] = Array[Byte](if (b) 1 else 0)
 
-  // fileModifiedBytes is a hot method, taking up 0.85% of reload time
-  // This is a procedural version
   def fileModifiedHash(f: File, digester: MessageDigest): Unit = {
-    if (f.isDirectory)
-      (f listFiles classDirFilter) foreach { x =>
-        fileModifiedHash(x, digester)
-      } else digester.update(bytes(getModifiedTimeOrZero(f)))
-
-    digester.update(bytes(f.getAbsolutePath))
+    Files.walkFileTree(
+      f.toPath,
+      new SimpleFileVisitor[Path] {
+        override def visitFile(file: Path, attrs: BasicFileAttributes) = {
+          if (classDirFilter accept file.toFile)
+            digester.update(bytes(attrs.lastModifiedTime.toMillis))
+          digester.update(bytes(file.toFile.getAbsolutePath))
```

Comment:
there should be a way to get the absolute path without going via File... see https://github.com/ensime/ensime-server/blob/3.0/api/src/main/scala/org/ensime/io/Canon.scala#L31-L42

Reply:
also, avoiding two `.toFile` calls would be good here

Reply:
is toFile expensive?

Reply:
`Path` uses an optimised hierarchical representation, whereas `File` constructs a full `String`

when I've profiled sbt on large projects, this has been the single biggest cause of heap usage.

Reply:
the alternative I found is `dir.toAbsolutePath.toString`

Reply:
it's a shame that you need to construct the string for the callback... I guess **that** optimisation needs to wait until the API can be changed.

Reply:
re-implemented!

Reply:
oh you mean the other string